go/enclave/components/rollup_compression.go

```diff
@@ -0,0 +1,385 @@
+	"github.com/ethereum/go-ethereum/trie"
+	"github.com/obscuronet/go-obscuro/go/common"
+	"github.com/obscuronet/go-obscuro/go/common/compression"
+	"github.com/obscuronet/go-obscuro/go/common/errutil"
+	"github.com/obscuronet/go-obscuro/go/common/log"
+	"github.com/obscuronet/go-obscuro/go/enclave/core"
+	"github.com/obscuronet/go-obscuro/go/enclave/crypto"
+	"github.com/obscuronet/go-obscuro/go/enclave/storage"
+)
+
+/*
+RollupCompression - responsible for the compression logic
+
+## Problem
+The main overhead (after the tx payloads), are the batch headers.
+
+## Requirements:
+1. recreate the exact batch headers as the live ones
+2. security - make sure it all chains up cryptographically, so it can't be gamed
+
+## Solution elements:
+1. Add another compressed and encrypted metadata blob to the ExtRollup. This is the "CalldataRollupHeader".
+It will also be published as calldata, together with the transactions. The role of this header is to contain the bare minimum
+information required to recreate the batches.
+2. Use implicit position information, deltas, and exceptions to minimise size.
+Eg. If the time between 2 batches is always 1second, there is no need to store any extra information.
+3. To avoid storing hashes, which don't compress at all, we execute each batch to be able to populate the parent hash.
+4. The Signatures over the batches are not stored, since the rollup is itself signed.
+5. The cross chain messages are calculated.
+*/
+type RollupCompression struct {
+	dataEncryptionService  crypto.DataEncryptionService
+	dataCompressionService compression.DataCompressionService
+	batchRegistry          BatchRegistry
+	batchExecutor          BatchExecutor
+	logger                 gethlog.Logger
+	storage                storage.Storage
+}
+
+func NewRollupCompression(
+	batchRegistry BatchRegistry,
+	batchExecutor BatchExecutor,
+	dataEncryptionService crypto.DataEncryptionService,
+	dataCompressionService compression.DataCompressionService,
+	storage storage.Storage,
+	logger gethlog.Logger,
+) *RollupCompression {
+	return &RollupCompression{
+		batchRegistry:          batchRegistry,
+		batchExecutor:          batchExecutor,
+		dataEncryptionService:  dataEncryptionService,
+		dataCompressionService: dataCompressionService,
+		logger:                 logger,
+		storage:                storage,
+	}
+}
+
+// temporary data structure to help build a batch from the information found in the rollup
+type batchFromRollup struct {
+	transactions []*common.L2Tx
+	seqNo        *big.Int
+	height       *big.Int
+	txHash       gethcommon.Hash
+	time         uint64
+	l1Proof      common.L1BlockHash
+
+	header *common.BatchHeader // for reorgs
+}
+
+// CreateExtRollup - creates a compressed and encrypted External rollup from the internal data structure
+func (rc *RollupCompression) CreateExtRollup(r *core.Rollup) (*common.ExtRollup, error) {
+	header, err := rc.createRollupHeader(r.Batches)
+	if err != nil {
+		return nil, err
```

Comment:
Would be good to wrap those errors in mroe descriptive ones; When running through errors in the logs it can be hard to find out during which bit of processing they happened.

Reply:
I know!
I was thinking we should use a library everywhere that does the wrapping and recreates some sort of sane stacktraces. I hate to manually wrap everything, and write some random repetitive text just to know where something happened.

Reply:
https://pkg.go.dev/github.com/pkg/errors - I think this one automatically creates stack traces

Reply:
let's do this as part of a different work. And add everywhere. 
We can use this library or another one